src/TokenStream.php

```diff
@@ -27,7 +27,11 @@ public function __construct(
         private array $tokens,
         private ?Source $source = null,
     ) {
-        $this->source = $source ?: new Source('', '');
+        if (null === $this->source) {
```

Comment:
This should change the property to a non-promoted one, so that the property type can be not-nullable

Reply:
I will do this in 4.0. As it's private, that doesn't change anything anyway, right (or maybe for static analyzer)?